annotatedcsv/decode.go

```diff
@@ -0,0 +1,451 @@
+	return nil
+}
+
+// fieldKind represents one of the possible kinds of struct field.
+// It's similar to reflect.Kind (every reflect.Kind constant i
+// is represented as fieldKind(i)) except that it also has
+// defined constants for field types that have special treatment,
+// such as time.Duration.
+type fieldKind uint
+
+// conv represents a possible conversion source and destination.
+type conv struct {
+	from colType
+	to   fieldKind
+}
+
+// intKinds holds all supported signed integer kinds
+var intKinds = []fieldKind{
+	fieldKind(reflect.Int),
+	fieldKind(reflect.Int8),
+	fieldKind(reflect.Int16),
+	fieldKind(reflect.Int32),
+	fieldKind(reflect.Int64),
+}
+
+// uintKinds holds all supported unsigned integer kinds
+var uintKinds = []fieldKind{
+	fieldKind(reflect.Uint),
+	fieldKind(reflect.Uint8),
+	fieldKind(reflect.Uint16),
+	fieldKind(reflect.Uint32),
+	fieldKind(reflect.Uint64),
+}
+
+// floatKinds holds all supported floating point number kinds
+var floatKinds = []fieldKind{
+	fieldKind(reflect.Float32),
+	fieldKind(reflect.Float64),
+}
+
+// rest of supported type kinds
+const (
+	durationKind = fieldKind(255 + iota)
+	timeKind
+	bytesKind
+)
+
+// colType is integer based representation annotated CSV types
+type colType uint
+
+const (
+	stringCol = colType(iota)
+	boolCol
+	durationCol
+	longCol
+	uLongCol
+	doubleCol
+	base64BinaryCol
+	timeColRFC
+	timeColRFCNano
+)
+
+// canonicalTypes holds the Go type that best represents
+// all the annotated CSV column kinds (keyed  by column type).
+var canonicalTypes = []reflect.Type{
+	stringCol:       reflect.TypeOf(""),
+	boolCol:         reflect.TypeOf(true),
+	durationCol:     reflect.TypeOf(time.Duration(1)),
+	longCol:         reflect.TypeOf(int64(0)),
+	uLongCol:        reflect.TypeOf(uint64(0)),
+	doubleCol:       reflect.TypeOf(0.0),
+	base64BinaryCol: reflect.TypeOf([]byte{}),
+	timeColRFC:      reflect.TypeOf(time.Time{}),
+	timeColRFCNano:  reflect.TypeOf(time.Time{}),
+}
+
+// columnTypes maps annotated CSV types
+// to integer column type
+var columnTypes = map[string]colType{
+	"string":               stringCol,
+	"boolean":              boolCol,
+	"duration":             durationCol,
+	"long":                 longCol,
+	"unsignedLong":         uLongCol,
+	"double":               doubleCol,
+	"base64Binary":         base64BinaryCol,
+	"dateTime:RFC3339":     timeColRFC,
+	"dateTime:RFC3339Nano": timeColRFCNano,
+	"dateTime":             timeColRFCNano,
+}
+
+// conversions maps all possible conversions from column types to field kinds
+var conversions map[conv]fieldSetter
+
+// fieldKindOf determines a fieldKind for given reflect.Type
+func fieldKindOf(t reflect.Type) fieldKind {
+	switch t {
+	case canonicalTypes[durationCol]:
+		return durationKind
+	case canonicalTypes[timeColRFC], canonicalTypes[timeColRFCNano]:
+		return timeKind
+	case canonicalTypes[base64BinaryCol]:
+		if t.Elem().Kind() == reflect.Uint8 {
+			return bytesKind
+		}
+	}
+	return fieldKind(t.Kind())
```

Comment:
This isn't quite right for interface types. We shouldn't be allowing non-empty interface types, but this logic conflates all interface types together, which means that `toInterface` will panic for non-empty interfaces.

```suggestion
        if t.Kind() == reflect.Interface && t != emptyInterfaceType {
                // Only the empty interface type is allowed.
                return reflect.Invalid
         }
	return fieldKind(t.Kind())
```

Let's add a test case for a non-empty interface too.

Reply:
Ok